aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -47,71 +41,128 @@ Gradle DSL. This presents a similar problem to `setup.py` in the Python world in
 that arbitrary logic can be executed during build time that could affect both
 how the customer's artifact is built, and which dependencies are chosen.
 
-To solve the issue where the build environment could affect the `build.gradle`
-script, a `--use-container` option will be provided, to give the option of
-building in an environment that closely mimics the Lambda environment.
+An interesting challenge is dealing with single build and multi build projects.
+Consider the following different projects structures:
+
+**Project A**
+```
+ProjectA
+├── build.gradle
+├── gradlew
+├── src
+└── template.yaml
+```
+
+**Project B**
+```
+ProjectB
+├── common
+│   └── build.gradle
+├── lambda1
+│   └── build.gradle
+├── lambda2
+│   └── build.gradle
+├── build.gradle
+├── gradlew
+├── settings.gradle
+└── template.yaml
+```
+
+Building Project A is relatively simple since we just need to issue `gradlew
+build` and place the built ZIP within the artifact directory.
+
+Building Project B is a little more complicated. Starting at the top of the
+directory, we begin by issuing `gradlew build` as before. However, since we have
+now built multiple functions, we need *multiple* artifact directories, one for
+each function. The build workflow must have some way of mapping each function's
+artifact to the correct artifact directory.
 
 ## Implementation
 
-We leverage Gradle to do all the heavy lifting for executing the `build.gradle`
-script which will resolve and download the dependencies and build the project.
-To create the distribution ZIP, we insert a new Gradle `task` that packages the
-compiled classes, resouces, and dependencies into a ZIP.
+### Source directory, and artifact directory semantics
+
+To enable the multi build projects where a Gradle project can contain multiple
+Lambdas as individual child projects, the builder behavior will change depending
+on whether the `options` provided to the build command contains a map called
+`artifact_mapping`.
+
+When this map is **not** present, the semantics of `source_dir` and
+`artifact_dir` are not changed.
+
+If this map is present, then `source_dir` is treated as the root directory of
+the parent project. Additionally, `artifact_dir` will not be the path under
+which the artifact will be copied to, but just the parent of the final
+directory. `artifact_mapping` will be a mapping from a source subdirectory under
```

Comment:
If the `artifact_mapping` is provided, should we optimize the build by doing 

```
for each mapping in artifact_mapping:
    cmd =+ " $mapping.folderName:build"

$GRADLE_EXECUTABLE --init-script /$SCRATCH_DIR/lambda-build-init.gradle $cmd
```

This does lead to issues where people make the sub-project name not match the folder (why people would I do not know...)
In that case we may have to make init script manipulare the task graph based on the project root of each sub-project. It's possible, just question around "should we"

Reply:
We can always optimize later if this is really needed.